Platform/TigerlakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c

```diff
@@ -520,6 +598,111 @@ AddSmbiosTypeString (
   return EFI_SUCCESS;
 }
 
+/**
+  Read Pch DTS
+
+**/
+VOID
+ReadPchDts (
+  VOID
+)
+{
+  UINT16          Dts;
+
+  // According to EDS, the register is in 2s complement, where
+  //   0x1: 1 deg C, 0x0: 0 deg C, 0x1FF: -1 deg C, and 0x1D8: -40 deg C
+  Dts = MmioRead16 (PCH_PWRM_BASE_ADDRESS + R_PMC_PWRM_THERMAL_TSS0) & B_PMC_PWRM_THERMAL_TSS0_TSR_MASK;
+  if (Dts <= 0xFF) {
+    mDtsInfo[DtsPch].NominalValue = Dts * 10;
+  } else {
+    // Prepend 0xFE00 because TSR bit fields are 9-bit
+    mDtsInfo[DtsPch].NominalValue = (Dts | 0xFE00) * 10;
+  }
+  mDtsInfo[DtsPch].LocationAndStatus.TemperatureProbeStatus = TemperatureProbeStatusOk;
+
+  DEBUG ((DEBUG_INFO, "Boot-time Pch Dts: 0x%x\n", mDtsInfo[DtsPch].NominalValue));
+}
+
+/**
+  Read CPU Dts
+
+**/
+VOID
+ReadCpuDts (
+  VOID
+)
+{
+  MSR_REGISTER    ThermStatus;
+  MSR_REGISTER    TemperatureTarget;
+  INT16           Dts;
+
+  ThermStatus.Qword = AsmReadMsr64(MSR_IA32_THERM_STATUS);
+
+  if ((ThermStatus.Dwords.Low & B_IA32_THERM_STATUS_VALID) == 0) {
+    return;
+  }
+
+  TemperatureTarget.Qword = AsmReadMsr64(MSR_TEMPERATURE_TARGET);
+  Dts = (INT16)TemperatureTarget.Bytes.ThirdByte - (INT16)ThermStatus.Bytes.ThirdByte;
+  mDtsInfo[DtsCpu].NominalValue = (UINT16)(Dts * 10);
+  mDtsInfo[DtsCpu].LocationAndStatus.TemperatureProbeStatus = TemperatureProbeStatusOk;
+
+  DEBUG ((DEBUG_INFO, "Boot-time Cpu DTS: 0x%x\n", mDtsInfo[DtsCpu].NominalValue));
+}
+
+/**
+  Append Boot DTS to Smbios
+
+  @retval EFI_SUCCESS             Initialized necessary information successfully
+  @retval EFI_OUT_OF_RESOURCES    Failed to retrieve stored Dts info
+  @retval EFI_OUT_OF_RESOURCES    Failed to allocate memory
+  @retval other                   other errors
+
+**/
+EFI_STATUS
+AppendSmbiosBootDts (
+  VOID
+  )
+{
+  EFI_STATUS        Status;
+  UINT16            Index;
+  UINT16            Cnt;
+  VOID              *TypeAddr;
+  UINT16            Length;
+  CHAR8             *StringPtr;
+
+  // allocate enough memory for type 28 and its string (only one string)
+  Length = sizeof (SMBIOS_TABLE_TYPE28) + SMBIOS_STRING_MAX_LENGTH + sizeof (CHAR8);
+  TypeAddr = AllocatePool (Length);
+  if (TypeAddr == NULL) {
+    DEBUG ((DEBUG_ERROR, "Unable to allocate memory\n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Cnt = sizeof (mDtsInfo) / sizeof (mDtsInfo[0]);
+  for (Index = 0; Index < Cnt; Index++) {
+    // copy type 28 structure
+    CopyMem (TypeAddr, &mDtsInfo[Index], sizeof (SMBIOS_TABLE_TYPE28));
+
+    // add 1st string: Description
+    StringPtr = (CHAR8 *)TypeAddr + sizeof (SMBIOS_TABLE_TYPE28);
+    StringPtr = AddSmbiosString (StringPtr, mDtsSite[Index]);
+
+    // additional null for last string
+    *StringPtr = 0;
+
+    // update total length
+    Length = (UINT16)(StringPtr - (CHAR8 *)TypeAddr + 1);
+
+    Status = AppendSmbiosType (TypeAddr, Length);
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+  }
+
+  return EFI_SUCCESS;
+}
+

```

Comment:
Stage2BoardInitLib.c is quite big. How about create a Dts.c to have these newly added dts related code?

Reply:
Sure. Dts.c and Dts.h are created.